pkg/rpc/client/helper.go

```diff
@@ -97,33 +106,107 @@ func topMapFromStack(st []stackitem.Item) (*stackitem.Map, error) {
+	// on the mainnet/testnet, see https://github.com/neo-project/neo-modules/pull/715#discussion_r897641254
+	// Thus, we'll create a short program that traverses iterator got from System.Contract.Call,
+	// packs all its values into array and stores it on stack.
+	script := io.NewBufBinWriter()
+	emit.Instruction(script.BinWriter, opcode.INITSLOT, // Initialize local slot...
+		[]byte{
+			2, // with 2 local variables (0-th for iterator, 1-th for the resulting array)...
+			0, // and 0 arguments.
+		})
+	// Pack arguments for System.Contract.Call.
+	if len(params) == 0 {
+		emit.Opcodes(script.BinWriter, opcode.NEWARRAY0)
+	} else {
+		bytes, err := json.Marshal(params)
+		if err != nil {
+			return nil, fmt.Errorf("failed to convert operation parameters into System.Contract.Call arguments: marshalling error %w", err)
+		}
+		var slice request.Params
+		err = json.Unmarshal(bytes, &slice)
+		if err != nil {
+			return nil, fmt.Errorf("failed to convert operation parameters into System.Contract.Call arguments: unmarshalling error %w", err)
+		}
+		err = request.ExpandArrayIntoScript(script.BinWriter, slice)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create function invocation script: %w", err)
+		}
+		emit.Int(script.BinWriter, int64(len(slice)))
+		emit.Opcodes(script.BinWriter, opcode.PACK)
+	}
+	emit.AppCallNoArgs(script.BinWriter, contract, operation, callflag.All) // The System.Contract.Call itself, it will push Iterator on estack.
+	emit.Opcodes(script.BinWriter, opcode.STLOC0,                           // Pop the result of System.Contract.Call (the iterator) from estack and store it inside the 0-th cell of the local slot.
+		opcode.NEWARRAY0, // Push new empty array to estack. This array will store iterator's elements.
+		opcode.STLOC1)    // Pop the empty array from estack and store it inside the 1-th cell of the local slot.
+
+	// Start the iterator traversal cycle.
+	iteratorTraverseCycleStartOffset := script.Len()
+	emit.Opcodes(script.BinWriter, opcode.LDLOC0)                   // Load iterator from the 0-th cell of the local slot and push it on estack.
+	emit.Syscall(script.BinWriter, interopnames.SystemIteratorNext) // Call System.Iterator.Next, it will pop the iterator from estack and push `true` or `false` to estack.
+	jmpIfNotOffset := script.Len()
+	emit.Instruction(script.BinWriter, opcode.JMPIFNOT, // Pop boolean value (from the previous step) from estack, if `false`, then iterator has no more items => jump to the end of program.
+		[]byte{
+			0x00, // jump to loadResultOffset, but we'll fill this byte after script creation.
+		})
+	emit.Opcodes(script.BinWriter, opcode.LDLOC1, // Load the resulting array from 1-th cell of local slot and push it to estack.
+		opcode.DUP, // Create a copy of the resulting array and push it to estack.
+		// TODO: this DUP actually is not needed if we remove STLOC1 a few lines below; array is reference type, it will be updated anyway.
```

Comment:
Done. But I have a doubt about it. I tried to write a short program in go that traverses iterator and packs its values into array.  The compiler generates these `DUP`s and after `APPEND` picks copy from stack, the initial array source located at the stack is being poped by `STLOC1`. Why the compiler emits these extra `DUP` and `STLOC1` opcodes? Seems that array is reference type, and array stored in the local slot will be updated anyway.

Reply:
Compiler emits `DUP` because `append(a, 1)` doesn't change `a` and thus we always use `a = append(a, 1)`. It could be a nice optimization in the compiler, though.

Reply:
Right, got it!